File: ios/Runner/MpvPlayer/MpvPlayerPlugin.swift
import UIKit
import Flutter

/// Flutter plugin that bridges MPV player to Dart via method and event channels
class MpvPlayerPlugin: NSObject, FlutterPlugin, FlutterStreamHandler, MpvPlayerDelegate {

    // MARK: - Properties

    private var playerCore: MpvPlayerCore?
    private var eventSink: FlutterEventSink?
    private weak var registrar: FlutterPluginRegistrar?

    // MARK: - FlutterPlugin Registration

    static func register(with registrar: FlutterPluginRegistrar) {
        // Method channel for commands
        let methodChannel = FlutterMethodChannel(
            name: "com.plezy/mpv_player",
            binaryMessenger: registrar.messenger()
        )

        // Event channel for state updates
        let eventChannel = FlutterEventChannel(
            name: "com.plezy/mpv_player/events",
            binaryMessenger: registrar.messenger()
        )

        let instance = MpvPlayerPlugin()
        instance.registrar = registrar

        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        eventChannel.setStreamHandler(instance)

        print("[MpvPlayerPlugin] Registered with Flutter")
    }

    // MARK: - FlutterStreamHandler

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        print("[MpvPlayerPlugin] Event stream connected")
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        print("[MpvPlayerPlugin] Event stream disconnected")
        return nil
    }

    // MARK: - FlutterPlugin Method Handler

    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initialize":
            handleInitialize(result: result)

        case "dispose":
            handleDispose(result: result)

        case "setProperty":
            handleSetProperty(call: call, result: result)

        case "getProperty":
            handleGetProperty(call: call, result: result)

        case "observeProperty":
            handleObserveProperty(call: call, result: result)

        case "command":
            handleCommand(call: call, result: result)

        case "setVisible":
            handleSetVisible(call: call, result: result)

        case "isInitialized":
            result(playerCore?.isInitialized ?? false)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Method Handlers

    private func handleInitialize(result: @escaping FlutterResult) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else {
                result(FlutterError(code: "ERROR", message: "Plugin deallocated", details: nil))
                return
            }

            // Check if already initialized
            if self.playerCore?.isInitialized == true {
                print("[MpvPlayerPlugin] Already initialized")
                result(true)
                return
            }

            // Find the key window
            guard let window = self.findKeyWindow() else {
                print("[MpvPlayerPlugin] Failed to find key window")
                result(FlutterError(code: "NO_WINDOW", message: "Could not find key window", details: nil))
                return
            }

            // Create and initialize player core
            let core = MpvPlayerCore()
            core.delegate = self

            guard core.initialize(in: window) else {
                print("[MpvPlayerPlugin] Failed to initialize MPV")
                result(FlutterError(code: "MPV_INIT_FAILED", message: "Failed to initialize MPV", details: nil))
                return
            }

            self.playerCore = core

            // Start hidden
            core.setVisible(false)

            print("[MpvPlayerPlugin] Initialized successfully")
            result(true)
        }
    }

    private func handleDispose(result: @escaping FlutterResult) {
        DispatchQueue.main.async { [weak self] in
            self?.playerCore?.dispose()
            self?.playerCore = nil
            print("[MpvPlayerPlugin] Disposed")
            result(nil)
        }
    }

    private func handleSetProperty(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let name = args["name"] as? String,
              let value = args["value"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing 'name' or 'value' argument", details: nil))
            return
        }

        playerCore?.setProperty(name, value: value)
        result(nil)
    }

    private func handleGetProperty(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let name = args["name"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing 'name' argument", details: nil))
            return
        }

        let value = playerCore?.getProperty(name)
        result(value)
    }

    private func handleObserveProperty(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let name = args["name"] as? String,
              let format = args["format"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing 'name' or 'format' argument", details: nil))
            return
        }

        playerCore?.observeProperty(name, format: format)
        result(nil)
    }

    private func handleCommand(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let commandArgs = args["args"] as? [String] else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing 'args' argument", details: nil))
            return
        }

        playerCore?.command(commandArgs)
        result(nil)
    }

    private func handleSetVisible(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let visible = args["visible"] as? Bool else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing 'visible' argument", details: nil))
            return
        }

        DispatchQueue.main.async { [weak self] in
            self?.playerCore?.setVisible(visible)

            // Update frame when becoming visible
            if visible {
                self?.playerCore?.updateFrame()
            }

            result(nil)
        }
    }

    // MARK: - MpvPlayerDelegate

    func onPropertyChange(name: String, value: Any?) {
        guard let eventSink = eventSink else { return }

        var event: [String: Any] = ["type": "property", "name": name]

        if let value = value {
            event["value"] = value
        }

        eventSink(event)
    }

    func onEvent(name: String, data: [String: Any]?) {
        guard let eventSink = eventSink else { return }

        var event: [String: Any] = ["type": "event", "name": name]
        if let data = data {
            event["data"] = data
        }

        eventSink(event)
    }

    // MARK: - Helpers

    private func findKeyWindow() -> UIWindow? {
        guard let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
              let window = windowScene.windows.first(where: { $0.isKeyWindow }) else {
            return nil
        }
        return window
    }
}

File: android/app/src/main/kotlin/com/edde746/plezy/mpv/MpvPlayerPlugin.kt
package com.edde746.plezy.mpv

import android.app.Activity
import android.util.Log
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class MpvPlayerPlugin : FlutterPlugin, MethodChannel.MethodCallHandler,
    EventChannel.StreamHandler, ActivityAware, MpvPlayerDelegate {

    companion object {
        private const val TAG = "MpvPlayerPlugin"
        private const val METHOD_CHANNEL = "com.plezy/mpv_player"
        private const val EVENT_CHANNEL = "com.plezy/mpv_player/events"
    }

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null
    private var playerCore: MpvPlayerCore? = null
    private var activity: Activity? = null
    private var activityBinding: ActivityPluginBinding? = null

    // FlutterPlugin

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel = MethodChannel(binding.binaryMessenger, METHOD_CHANNEL)
        methodChannel.setMethodCallHandler(this)

        eventChannel = EventChannel(binding.binaryMessenger, EVENT_CHANNEL)
        eventChannel.setStreamHandler(this)

        Log.d(TAG, "Attached to engine")
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        Log.d(TAG, "Detached from engine")
    }

    // ActivityAware

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        activityBinding = binding
        binding.addOnUserLeaveHintListener { playerCore?.onPause() }
        Log.d(TAG, "Attached to activity")
    }

    override fun onDetachedFromActivity() {
        playerCore?.dispose()
        playerCore = null
        activity = null
        activityBinding = null
        Log.d(TAG, "Detached from activity")
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
        activityBinding = binding
        Log.d(TAG, "Reattached to activity for config changes")
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
        activityBinding = null
        Log.d(TAG, "Detached from activity for config changes")
    }

    // EventChannel.StreamHandler

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
        Log.d(TAG, "Event stream connected")
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
        Log.d(TAG, "Event stream disconnected")
    }

    // MethodChannel.MethodCallHandler

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "initialize" -> handleInitialize(result)
            "dispose" -> handleDispose(result)
            "setProperty" -> handleSetProperty(call, result)
            "getProperty" -> handleGetProperty(call, result)
            "observeProperty" -> handleObserveProperty(call, result)
            "command" -> handleCommand(call, result)
            "setVisible" -> handleSetVisible(call, result)
            "isInitialized" -> result.success(playerCore?.isInitialized ?: false)
            else -> result.notImplemented()
        }
    }

    private fun handleInitialize(result: MethodChannel.Result) {
        val currentActivity = activity
        if (currentActivity == null) {
            result.error("NO_ACTIVITY", "Activity not available", null)
            return
        }

        if (playerCore?.isInitialized == true) {
            Log.d(TAG, "Already initialized")
            result.success(true)
            return
        }

        currentActivity.runOnUiThread {
            try {
                playerCore = MpvPlayerCore(currentActivity).apply {
                    delegate = this@MpvPlayerPlugin
                }
                val success = playerCore?.initialize() ?: false

                // Start hidden
                playerCore?.setVisible(false)

                Log.d(TAG, "Initialized: $success")
                result.success(success)
            } catch (e: Exception) {
                Log.e(TAG, "Failed to initialize: ${e.message}", e)
                result.error("INIT_FAILED", e.message, null)
            }
        }
    }

    private fun handleDispose(result: MethodChannel.Result) {
        activity?.runOnUiThread {
            playerCore?.dispose()
            playerCore = null
            Log.d(TAG, "Disposed")
            result.success(null)
        } ?: result.success(null)
    }

    private fun handleSetProperty(call: MethodCall, result: MethodChannel.Result) {
        val name = call.argument<String>("name")
        val value = call.argument<String>("value")

        if (name == null || value == null) {
            result.error("INVALID_ARGS", "Missing 'name' or 'value'", null)
            return
        }

        playerCore?.setProperty(name, value)
        result.success(null)
    }

    private fun handleGetProperty(call: MethodCall, result: MethodChannel.Result) {
        val name = call.argument<String>("name")

        if (name == null) {
            result.error("INVALID_ARGS", "Missing 'name'", null)
            return
        }

        val value = playerCore?.getProperty(name)
        result.success(value)
    }

    private fun handleObserveProperty(call: MethodCall, result: MethodChannel.Result) {
        val name = call.argument<String>("name")
        val format = call.argument<String>("format")

        if (name == null || format == null) {
            result.error("INVALID_ARGS", "Missing 'name' or 'format'", null)
            return
        }

        playerCore?.observeProperty(name, format)
        result.success(null)
    }

    private fun handleCommand(call: MethodCall, result: MethodChannel.Result) {
        val args = call.argument<List<String>>("args")

        if (args == null) {
            result.error("INVALID_ARGS", "Missing 'args'", null)
            return
        }

        playerCore?.command(args.toTypedArray())
        result.success(null)
    }

    private fun handleSetVisible(call: MethodCall, result: MethodChannel.Result) {
        val visible = call.argument<Boolean>("visible")

        if (visible == null) {
            result.error("INVALID_ARGS", "Missing 'visible'", null)
            return
        }

        playerCore?.setVisible(visible)
        result.success(null)
    }

    // MpvPlayerDelegate

    override fun onPropertyChange(name: String, value: Any?) {
        eventSink?.success(
            mapOf(
                "type" to "property",
                "name" to name,
                "value" to value
            )
        )
    }

    override fun onEvent(name: String, data: Map<String, Any>?) {
        val event = mutableMapOf<String, Any>(
            "type" to "event",
            "name" to name
        )
        data?.let { event["data"] = it }
        eventSink?.success(event)
    }
}
